File: lib/app/app_container.dart
import '../ai/gemini_ai_repository.dart';
import '../core/data/book_repository_impl.dart';
import '../core/data/importer/book_import_metadata_extractor.dart';
import '../core/data/importer/book_metadata_extractor.dart';
import '../core/database/app_database.dart';
import '../core/database/book_dao.dart';
import '../core/reader/epub/epub_reader_engine.dart';
import '../core/reader/pdf_reader_engine.dart';
import '../core/reader/reader_engine_registry.dart';
import '../core/storage/app_private_book_file_storage.dart';
import '../core/storage/book_file_storage.dart';
import '../domain/repository/ai_repository.dart';
import '../domain/repository/book_repository.dart';
import '../domain/usecase/get_library_books_use_case.dart';
import '../domain/usecase/import_book_use_case.dart';
import '../domain/usecase/open_book_use_case.dart';
import '../domain/usecase/save_reading_progress_use_case.dart';
import '../domain/usecase/search_in_book_use_case.dart';
import '../feature/reader/prefs/reader_preferences_store.dart';

class AppContainer {
  late final BookDao _bookDao;
  late final BookFileStorage _fileStorage;
  late final BookMetadataExtractor _metadataExtractor;
  late final BookRepository bookRepository;
  late final AiRepository aiRepository;
  late final ReaderEngineRegistry readerEngineRegistry;
  late final ReaderPreferencesStore readerPreferencesStore;

  late final GetLibraryBooksUseCase getLibraryBooksUseCase;
  late final ImportBookUseCase importBookUseCase;
  late final OpenBookUseCase openBookUseCase;
  late final SaveReadingProgressUseCase saveReadingProgressUseCase;
  late final SearchInBookUseCase searchInBookUseCase;

  Future<void> initialize() async {
    final db = await AppDatabase.getInstance();
    _bookDao = BookDao(db);
    _fileStorage = AppPrivateBookFileStorage();
    _metadataExtractor = BookImportMetadataExtractor();

    bookRepository = BookRepositoryImpl(
      bookDao: _bookDao,
      fileStorage: _fileStorage,
      metadataExtractor: _metadataExtractor,
    );

    aiRepository = GeminiAiRepository();

    readerEngineRegistry = ReaderEngineRegistry([
      EpubReaderEngine(),
      PdfReaderEngine(),
    ]);

    readerPreferencesStore = ReaderPreferencesStore();
    await readerPreferencesStore.load();

    getLibraryBooksUseCase = GetLibraryBooksUseCase(bookRepository);
    importBookUseCase = ImportBookUseCase(bookRepository);
    openBookUseCase = OpenBookUseCase(bookRepository);
    saveReadingProgressUseCase = SaveReadingProgressUseCase(bookRepository);
    searchInBookUseCase = SearchInBookUseCase(bookRepository);
  }

  BookFileStorage get bookFileStorage => _fileStorage;
}

File: app/src/main/kotlin/com/readler/app/AppContainer.kt
package com.readler.app

import android.content.Context
import com.readler.ai.GeminiAiRepository
import com.readler.app.telemetry.FileTelemetryLogger
import com.readler.app.telemetry.TelemetryLogger
import com.readler.core.data.BookRepositoryImpl
import com.readler.core.data.importer.BookImportMetadataExtractor
import com.readler.core.data.importer.BookMetadataExtractor
import com.readler.core.database.AppDatabase
import com.readler.core.reader.EpubReaderEngine
import com.readler.core.reader.PdfReaderEngine
import com.readler.core.reader.ReaderEngineRegistry
import com.readler.core.storage.AppPrivateBookFileStorage
import com.readler.domain.repository.BookRepository
import com.readler.domain.repository.AiRepository
import com.readler.domain.usecase.GetLibraryBooksUseCase
import com.readler.domain.usecase.ImportBookUseCase
import com.readler.domain.usecase.OpenBookUseCase
import com.readler.domain.usecase.SearchInBookUseCase
import com.readler.domain.usecase.SaveReadingProgressUseCase
import com.readler.feature.reader.prefs.SharedReaderPreferencesStore

object AppContainer {
    private lateinit var appContext: Context

    fun initialize(context: Context) {
        if (!::appContext.isInitialized) {
            appContext = context.applicationContext
        }
    }

    private val database: AppDatabase by lazy { AppDatabase.getInstance(appContext) }
    private val fileStorage by lazy { AppPrivateBookFileStorage(appContext) }
    private val metadataExtractor: BookMetadataExtractor by lazy { BookImportMetadataExtractor(appContext) }
    val telemetryLogger: TelemetryLogger by lazy { FileTelemetryLogger(appContext) }
    val aiRepository: AiRepository by lazy {
        GeminiAiRepository(
            context = appContext,
            defaultApiKey = BuildConfig.GEMINI_API_KEY
        )
    }

    val bookRepository: BookRepository by lazy {
        BookRepositoryImpl(
            bookDao = database.bookDao(),
            fileStorage = fileStorage,
            metadataExtractor = metadataExtractor
        )
    }

    val getLibraryBooksUseCase by lazy { GetLibraryBooksUseCase(bookRepository) }
    val importBookUseCase by lazy { ImportBookUseCase(bookRepository) }
    val openBookUseCase by lazy { OpenBookUseCase(bookRepository) }
    val saveReadingProgressUseCase by lazy { SaveReadingProgressUseCase(bookRepository) }
    val searchInBookUseCase by lazy { SearchInBookUseCase(bookRepository) }

    val readerEngineRegistry by lazy {
        ReaderEngineRegistry(
            readerEngines = listOf(
                EpubReaderEngine(),
                PdfReaderEngine()
            )
        )
    }

    val bookFileStorage by lazy { fileStorage }
    val readerPreferencesStore by lazy { SharedReaderPreferencesStore(appContext) }
}
